mesonbuild/meson
# Major rewrite of the rewriter and the static introspection tool

The rewriter and the static introspection tool used to be very broken, now it is *less* broken.

The most important changes are:

1. We now have `class UnknownValue` for more explicit handling of situations that are too complex/impossible.
2. If you write
```
var = 'foo'
name = var
var = 'bar'
executable(name, 'foo.c')
```
the tool now knows that the name of the executable is `foo` and not `bar`. See `dataflow_dag` and `node_to_runtime_value` for details on how we do this.

To test my work I wrote a script that:
1. `git clone`'s a couple of big projects using meson (e.g. systemd).
3. Checks if the outputs of `meson introspect meson.build --targets` and `meson introspect build_folder --targets` are not contradicting each other.
4. Checks if the output of `meson introspect meson.build --targets` is the same for two different versions of meson (the one we want to test and a known good version).
5. Iterates over all targets that are found using  `meson introspect meson.build --targets` and checks if
```
meson rewrite target tgt add rewrite_test_source.c
meson rewrite target tgt add rewrite_test_source.c
meson introspect meson.build --targets
meson rewrite target tgt rm rewrite_test_source.c
meson introspect meson.build --targets
```
does not crash and produces the expected output.

I think this script is very useful (it found a ton of bugs), but I do not know where to put it, so it currently only exists on my machine. It is too slow (1 hour iirc, haven't measured) to run it in the CI pipeline. [In the docs](https://mesonbuild.com/Users.html) you write
```
All the software on this list is tested for regressions before release
```
Is this testing (partially) automated? If so, could we merge my script with it?


@bruchar1 @kcgen Afaik you are one of the few people using the rewriter/static introspection tool in production. Could you test your usecases and voice your opinion?

@eli-schwartz You promised me this in a mail:

> I would be willing to review the resulting PRs with the intent of
checking that the results "seem to work" and not getting bogged down in
nitty-gritty details. 🙂 I agree that it's relatively obscure
functionality that doesn't always work and the priority should be,
essentially, treating it as brand new code.
